ci: move the remaining travis tests to circle

IgorMinar opened the pull request:
This is the last chunk of work to move us off of Travis.

Huge thanks to @devversion for helping me work through this PR.

IgorMinar committed:
ci: move local and saucelabs unit tests to circle

Moving the tests over to CircleCI in pretty much "as-is" state just so that we can drop the dependency on Travis.

In the followup changes we plan to migrate these tests to run on sauce under bazel. @gregmagolan is working on that.

I've previously verified that all the tests executed in legacy-unit-tests-local already under bazel.
Therefore the legacy-unit-tests-local job is strictly not necessary any more, but given how flaky legacy-unit-tests-saucelabs is,
it is good to have the -local job just so that we can quickly determine if any failure is a flake or legit issue
(the bazel version of these tests could theoretically run in a slightly different way and fail or not fail in a different way, so having -lcoal job is just an extra safety check).

This change was coauthored with @devversion

mary-poppins commented:
You can preview 227c52e at https://pr27937-227c52e.ngbuilds.io/.

IgorMinar commented on scripts/sauce/sauce_connect_wait_for_tunnel.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
I'm not even sure if we can use this script now that the build happens in a different job...

IgorMinar committed:
ci: remove build steps that are no longer needed

the metadata build seems to be needed only by the offline compiler tests which is currently disabled

mary-poppins commented:
You can preview 23320fc at https://pr27937-23320fc.ngbuilds.io/.

mary-poppins commented:
You can preview 5e5db78 at https://pr27937-5e5db78.ngbuilds.io/.

mary-poppins commented:
You can preview cbd3a3d at https://pr27937-cbd3a3d.ngbuilds.io/.

mary-poppins commented:
You can preview bf9e746 at https://pr27937-bf9e746.ngbuilds.io/.

mary-poppins commented:
You can preview 0c566fc at https://pr27937-0c566fc.ngbuilds.io/.

mary-poppins commented:
You can preview 9d78561 at https://pr27937-9d78561.ngbuilds.io/.

mary-poppins commented:
You can preview f026b19 at https://pr27937-f026b19.ngbuilds.io/.

mary-poppins commented:
You can preview df5edde at https://pr27937-df5edde.ngbuilds.io/.

mary-poppins commented:
You can preview fcda67b at https://pr27937-fcda67b.ngbuilds.io/.

mary-poppins commented:
You can preview 39557e7 at https://pr27937-39557e7.ngbuilds.io/.

mary-poppins commented:
You can preview 9e9d4fa at https://pr27937-9e9d4fa.ngbuilds.io/.

devversion committed:
ci: move e2e tests from travis to circleci

mary-poppins commented:
You can preview feb22da at https://pr27937-feb22da.ngbuilds.io/.

mary-poppins commented:
You can preview a07af54 at https://pr27937-a07af54.ngbuilds.io/.

mary-poppins commented:
You can preview 8f17072 at https://pr27937-8f17072.ngbuilds.io/.

mary-poppins commented:
You can preview e2eaa3c at https://pr27937-e2eaa3c.ngbuilds.io/.

mary-poppins commented:
You can preview 0a6cd30 at https://pr27937-0a6cd30.ngbuilds.io/.

mary-poppins commented:
You can preview f5446d9 at https://pr27937-f5446d9.ngbuilds.io/.

mary-poppins commented:
You can preview 4149e86 at https://pr27937-4149e86.ngbuilds.io/.

mary-poppins commented:
You can preview 086ae03 at https://pr27937-086ae03.ngbuilds.io/.

mary-poppins commented:
You can preview e20e272 at https://pr27937-e20e272.ngbuilds.io/.

IgorMinar commented on .circleci/config.yml:
```diff
@@ -468,14 +468,26 @@ jobs:
       - run:
           name: Building Angular benchpress output
           command: |
+            bazelBinary=$(yarn bin bazel)
```
this variable seems to decrease readability - it's used only one in a nested expansion.

I'd prefer the following:

IgorMinar commented on .circleci/config.yml:
```diff
@@ -468,14 +468,26 @@ jobs:
           name: Building Angular benchpress output
           command: |
+            bazelBinary=$(yarn bin bazel)
+            bazelBinOutput=$(${bazelBinary} info bazel-bin)
```
bazelBinDir=$(yarn bazel info bazel-bin)

IgorMinar commented on .circleci/config.yml:
```diff
@@ -468,14 +468,26 @@ jobs:
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-examples-e2e.conf.js --bundles=true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-perf.conf.js --bundles=true --dryrun
+
+      - run: yarn gulp source-map-test
```
this one doesn't really have to do anything with e2e tests. maybe we should create a new job called `legacy-misc-tests` and move this one and cycle check in there...

devversion replied:
Good point.

IgorMinar commented on modules/tsconfig.json:
```diff
@@ -12,7 +12,7 @@
       "selenium-webdriver": ["../node_modules/@types/selenium-webdriver/index.d.ts"],
       "rxjs/*": ["../node_modules/rxjs/*"],
-      "@angular/*": ["../dist/all/@angular/*"]
+      "@angular/*": ["../dist/packages-dist/*"]
```
👍

IgorMinar commented on tools/gulp-tasks/check-cycle.js:
```diff
@@ -10,7 +10,10 @@
-  const dependencyObject = madge(['dist/all/'], {
+  // TODO: This only checks for circular dependencies within each package because
+  // imports to other packages cannot be resolved by Madge when CommonJS is used.
+  // We should consider updating Madge and use a tsonfig to check across packages.
```
how did you discover this limitation? 

if this is true, then we should prioritize moving this test over to bazel, and in bazel creating a package-less file layout similar to `dist/all` and run the check there. (or updating madge, whatever is easier).

the reason for this tests is that in google3 we can't have cycles due to closure compiler, so it is important that we don't introduce any otherwise g3sync will fail. it's not common for us to introduce cycle especially across packages, so it's ok to degrade this test for a bit.

devversion replied:
I just ran Madge manually and it looked like it wasn't able to resolve imports like `@angular/core` to the proper entry point. We never specified any path mappings, so I'm fairly confident that it doesn't do that right now.

Running like `yarn madge --format cjs --extension js --summary --break-on-error dist/all` causes it to break while resolving `@angular/animations`. We should look into this, agreed, but that seems unrelated to this change, and would be better as a follow-up.

IgorMinar commented on modules/benchmarks/e2e_test/tree_data.ts:
```diff
@@ -36,8 +36,6 @@ export const Benchmarks: Benchmark[] = [
     url: 'all/benchmarks/src/tree/ng2_next/index.html',
     buttons: CreateDestroyDetectChangesButtons,
     ignoreBrowserSynchronization: true,
-    // Can't use bundles as we use non exported code
```
this is no longer true. so it's ok to use bundles.

IgorMinar commented on modules/tsconfig.json:
```diff
@@ -12,7 +12,7 @@
       "selenium-webdriver": ["../node_modules/@types/selenium-webdriver/index.d.ts"],
       "rxjs/*": ["../node_modules/rxjs/*"],
-      "@angular/*": ["../dist/packages-dist/*"]
+      "@angular/*": ["../dist/all/@angular/*"]
```
why did you revert this change? I liked the way you had it. there must have been a reason but you didn't mention it in the commit message.

devversion committed:
ci: run legacy e2e tests in parallel

mary-poppins commented:
You can preview 438b061 at https://pr27937-438b061.ngbuilds.io/.

mary-poppins commented:
You can preview 9435c12 at https://pr27937-9435c12.ngbuilds.io/.

devversion committed:
ci: remove travis ci setup

we no longer need it... yay!!!

mary-poppins commented:
You can preview e43bd28 at https://pr27937-e43bd28.ngbuilds.io/.

devversion committed:
build: remove "build.sh" script

this script is now obsolete and not needed any more. yay!!!

devversion committed:
build: remove "test.sh" script

test.sh is no longer needed... all the tests should now be executed via bazel.

if for whatever reason we need to run the legacy unit test setup, we should should follow the commands that we use to execute those tests in .circle/config.yaml

mary-poppins commented:
You can preview 8dae96e at https://pr27937-8dae96e.ngbuilds.io/.

mary-poppins commented:
You can preview 1bff248 at https://pr27937-1bff248.ngbuilds.io/.

mary-poppins commented:
You can preview e6ed5b1 at https://pr27937-e6ed5b1.ngbuilds.io/.

IgorMinar committed:
test: fix integration/platform-server test which had missing @types/node devDep

I'm not sure why this problem is visible only now or how this worked before, but the CI
is now failing because @types/node is missing.

I also added the yarn.lock file which was previously omitted. We want the yarn.lock file in so that
our deps don't change over time without us knowing.

mary-poppins commented:
You can preview 697c794 at https://pr27937-697c794.ngbuilds.io/.

mary-poppins commented:
You can preview f0f6bf0 at https://pr27937-f0f6bf0.ngbuilds.io/.

mary-poppins commented:
You can preview ac7540e at https://pr27937-ac7540e.ngbuilds.io/.

mary-poppins commented:
You can preview 4968a5f at https://pr27937-4968a5f.ngbuilds.io/.

mary-poppins commented:
You can preview 0e19549 at https://pr27937-0e19549.ngbuilds.io/.

mary-poppins commented:
You can preview 8eee4ce at https://pr27937-8eee4ce.ngbuilds.io/.

mary-poppins commented:
You can preview d1f3e16 at https://pr27937-d1f3e16.ngbuilds.io/.

IgorMinar commented on .circleci/config.yml:
```diff
@@ -462,3 +562,14 @@ workflows:
+# - add folding
+# - check local chrome version pulled from docker image
+# - remove /tools/ngcontainer
+# - could we install sauce-connect via yarn?
```
done

IgorMinar commented on .circleci/config.yml:
```diff
@@ -462,3 +562,14 @@ workflows:
+# - verify that we are bootstrapping with the right yarn version coming from the docker image
+# - do we need node_modules check?
+#   (node tools/npm/check-node-modules --purge && yarn postinstall) || yarn install --frozen-lockfile --non-interactive
+# - add folding
```
not needed

IgorMinar commented on .circleci/config.yml:
```diff
@@ -462,3 +562,14 @@ workflows:
+# - check local chrome version pulled from docker image
+# - remove /tools/ngcontainer
+# - could we install sauce-connect via yarn?
+# - could we have one circleci vm that would be the saucelabs tunnel for the entire workflow?
```
not needed here, but something we might want to consider for running the unit tests under bazel...

IgorMinar commented:
this PR does not affect g3: https://test.corp.google.com/ui#id=OCL:228041852:BASE:228041924:1546769193258:89b97bbd

devversion committed:
build: shard integration tests on circleci

mary-poppins commented:
You can preview 201ce9c at https://pr27937-201ce9c.ngbuilds.io/.

mary-poppins commented:
You can preview 842fc64 at https://pr27937-842fc64.ngbuilds.io/.

devversion committed:
build: bazel integration test using limited resources

mary-poppins commented:
You can preview dbdae48 at https://pr27937-dbdae48.ngbuilds.io/.

mary-poppins commented:
You can preview 1357290 at https://pr27937-1357290.ngbuilds.io/.

mary-poppins commented:
You can preview a461d9d at https://pr27937-a461d9d.ngbuilds.io/.

devversion committed:
build: group sharded "test_docs_examples" jobs within circleci

* Groups the two sharded `test_docs_examples` job using CircleCI's `parallelism` feature.  This makes the amount of jobs that show up on a PR, more reduced and also reduces code duplication for maintaining the Circle job definition.

mary-poppins commented:
You can preview 343592c at https://pr27937-343592c.ngbuilds.io/.

mary-poppins commented:
You can preview a6ea4ba at https://pr27937-a6ea4ba.ngbuilds.io/.

gkalpak commented on .circleci/config.yml:
```diff
@@ -218,11 +218,12 @@ jobs:
     docker:
       # Needed because the example e2e tests depend on Chrome.
       - image: *browsers_docker_image
+    parallelism: 2
```
With all the changes in this PR, this is now the slowest path. Increasing parallelism here could save use another 60-90 secs :wink:)

IgorMinar replied:
we can give it a shot. long term all these tests should run via bazel on RBE and we should not need to use circle parallelism.

one thing to keep in mind is that a big downside of circle's paralellism feature is that it's easy to incorrectly allocate resources and either under-allocate or over-allocate VMs. with bazel and rbe we are getting automatic resource allocation that doesn't need to be closely managed and monitored.

devversion replied:
Agreed. I've just switched it to three parallel instances for now.

gkalpak reviewed (dismissed):
Woohoooo!!! 🎉 🎉 🎉 

Letf a few minor comments/suggestions, but generally LGTM 💯 
A couple more comments:
- It might be good enough for now, but the `integration_tests` sharding can be further improved to get more uniform times per container. Currently we get: 6m, 7m, 3.5m, 5.5m.
- This PR also fixes #21707 and #27663. Please, add that info in the appropriate commits.

IgorMinar commented on modules/build.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
+# Immediately exit if any command failed.
+set -e
```
yes. please use the same prefix as what we use everywhere:

```
set -x -u -e -o pipefail
```

IgorMinar commented on modules/build.sh:
```diff
@@ -0,0 +1,47 @@
+set -e
+
+# Go to project directory.
+cd $(dirname ${0})/..
```
what other script does this wrapping? I'm not aware.

typically we do the following:

https://github.com/angular/angular/blob/master/scripts/ci/build.sh#L6

and we use the var to resolve paths and not rely on CWD

gkalpak replied:
> what other script does this wrapping? I'm not aware.

@IgorMinar, you introduced this "concept" in b8f0c3dc7 ([example 1][1], [example 2][2], [example 3][3]).
But maybe my understanding of the reasoning is incorrect.

Just to be clear, I was referring to changing this:

```
cd some/dir/
run some code
```

to that:

```
(
  cd some/dir/
  run some code
)
```

##
@devversion, good point about `$0`. It will indeed not work as expected when sourced.
I am either misunderstanding the point of wrapping in `(...)` or there is no point here anyway.

In any case, I don't feel strongly at all about.

[1]: https://github.com/angular/angular/commit/b8f0c3dc7bdf37553fa564578635f5bfcc6c646c#diff-cf6136203c5f24fba45e4ab92e7895e3R52
[2]: https://github.com/angular/angular/commit/b8f0c3dc7bdf37553fa564578635f5bfcc6c646c#diff-241beae81a9a4b50c44ecde82e496324R11
[3]: https://github.com/angular/angular/commit/b8f0c3dc7bdf37553fa564578635f5bfcc6c646c#diff-5ead0d120428a4b905965643dac1af0eR11

devversion replied:
Using the `(...)` basically makes this execute in a subshell, so that `cd` does not have an effect on the primary shell. 

Though since we don't support running this script with `source`, the directory change will never affect the main shell because the "normal running" of the script also runs in a subshell.

IgorMinar replied:
@gkalpak right. I did that, but it only makes sense AFAIK if you cd as part of a portion of a script. doing it at the beginning of a script doesn't really make a difference because CWD will be reset after the script exists.

gkalpak replied:
Right. So, it only makes sense if you want to source the script or have multiple `cd`s in the same file (neither of which is true here).
Thanks for clarifying 👍

IgorMinar commented on integration/get-sharded-tests.js:
```diff
@@ -0,0 +1,96 @@
+// Current node index of the CircleCI job. A CircleCI job can contain multiple nodes if the
+// parallelism option is enabled. This value is used to determine the tests that need to be
+// run for this instance.
+const containerInstanceIndex = Number(process.env['CIRCLE_NODE_INDEX']);
```
SGTM

IgorMinar commented on integration/get-sharded-tests.js:
```diff
@@ -0,0 +1,96 @@
+  let assignedTests = 0;
+
+  for (let i = 0; i < maxInstances; i++) {
+    const shardSize = Math.floor((tests.length - assignedTests) / (maxInstances - i));
```
all shards start in parallel as far as I can tell so `ceil` will have no significant impact.

IgorMinar replied:
@gkalpak module approach doesn't take into account that some integration tests take much longer and need their own special shard. I think Pauls current approach is good and eventually we just need to replace all of this with bazel+rbe and removing the sharding.

devversion replied:
> all shards start in parallel as far as I can tell so ceil will have no significant impact.

Agreed on that point. Though if we make this "CI-agnostic", it would be better to use "ceil" in case there is a container queue.

>  module approach doesn't take into account that some integration tests take much longer and need their own special shard. I think Pauls current approach is good and eventually we just need to replace all of this with bazel+rbe and removing the sharding.

I initially thought the same and that's why I didn't use the `--shards {moduloResult}/{moduloDivider}` approach, but I think I got it working with modulo and it feels more clean and less verbose.

IgorMinar commented on docs/DEVELOPER.md:
````diff
@@ -82,7 +82,7 @@ Before submitting a PR, do not forget to remove them:
 
 ```shell
-./build.sh
+./scripts/build-packages-dist.sh
````
+1

devversion replied:
That's true, but what would be the proper message if we only remove `build.sh`? (if we only consider 2431a97e40793ca65dee6094fc8777819d13b662)

I will just swap the commits, so that we remove `test.sh` before `build.sh`. It should be fine then.

devversion commented on integration/get-sharded-tests.js:
```diff
@@ -0,0 +1,96 @@
+
+// List of all integration tests that are available.
+const integrationTests =
+    fs.readdirSync(path.join(__dirname))
```
Good catch 😄

devversion commented on scripts/saucelabs/wait-for-tunnel.sh:
```diff
@@ -0,0 +1,24 @@
+
+  # Counter needs to be multiplied by two because the while loop only sleeps a half second.
+  # This has been made in favor of better progress logging (printing dots every half second)
+  if [ $counter -gt $[${SAUCE_READY_FILE_TIMEOUT} * 2] ]; then
```
I don't think that changes too often and rather makes the script more complicated than it needs to be.

devversion commented on scripts/saucelabs/wait-for-tunnel.sh:
```diff
@@ -0,0 +1,24 @@
+  # This has been made in favor of better progress logging (printing dots every half second)
+  if [ $counter -gt $[${SAUCE_READY_FILE_TIMEOUT} * 2] ]; then
+    echo ""
+    echo "Timed out after 2 minutes waiting for tunnel ready file"
```
Good idea

devversion commented on .circleci/config.yml:
```diff
@@ -444,6 +444,48 @@ jobs:
+          background: true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-e2e.conf.js --bundles=true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-examples-e2e.conf.js --bundles=true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-perf.conf.js --bundles=true --dryrun
```
It technically does. Running with yarn usually brings a modified shell environment. Not sure if this is a problem here, so I'll try running with `yarn`.

gkalpak replied:
If there is a specific reason not to use `yarn protractor`, I would consider leaving a comment explaining that.
I would also use `yarn bin` over `npm bin`, because we are generally using yarn :wink:

devversion commented on .circleci/config.yml:
```diff
@@ -444,6 +444,48 @@ jobs:
+      - run: yarn gulp check-cycle
+      # TODO: disabled because the Bazel packages-dist does not seem to have map files for
+      # the ESM5/ES2015 output.
+      # - run: yarn gulp source-map-test
```
There wasn't. I've just created a Github issue for tracking. https://github.com/angular/angular-cli/issues/13378

devversion commented on .circleci/env.sh:
```diff
@@ -31,5 +34,26 @@ setSecretVar CI_SECRET_PAYLOAD_FIREBASE_TOKEN "$ANGULAR_PAYLOAD_TOKEN";
+# In order to have a meaningful SauceLabs badge on the repo page,
+# the angular2-ci account is used only when pushing commits to master;
+# in all other cases, the regular angular-ci account is used.
+if [ "${CI_PULL_REQUEST}" = "false" ] && [ "${CI_BRANCH}" = "master" ]; then
```
This should be guarded by the `CI_PULL_REQUEST` check. Additionally if Igor makes the access key a secret CircleCI variable, we should be guarded as well.

devversion replied:
True, it would guard against that if they don't update their setup. Best would be if Igor makes this is a secret CircleCI key. I'll add the safety check according to your feedback anyway 👍

devversion commented on .circleci/env.sh:
```diff
@@ -31,5 +34,26 @@ setSecretVar CI_SECRET_PAYLOAD_FIREBASE_TOKEN "$ANGULAR_PAYLOAD_TOKEN";
+# in all other cases, the regular angular-ci account is used.
+if [ "${CI_PULL_REQUEST}" = "false" ] && [ "${CI_BRANCH}" = "master" ]; then
+  setPublicVar SAUCE_USERNAME "angular2-ci";
+  setSecretVar SAUCE_ACCESS_KEY "693ebc16208a-0b5b-1614-8d66-a2662f4e";
```
cc. @IgorMinar. I think this is reasonable. This would *ensure* that the Saucelabs badge is 100% based on the results of the `angular/angular#master` branch.

devversion commented on docs/DEVELOPER.md:
```diff
@@ -85,30 +85,20 @@ To build Angular run:
-- CircleCI fails if your code is not formatted properly,
-- Travis CI fails if any of the test suites described above fails.
+All the tests are executed on our Continuous Integration infrastructure and a PR could only be
+merged once all tests pass. CircleCI fails if your code is not is passing all tests or is not
```
Agreed. I basically just based it on what was there before 😄 CircleCI and Travis were mentioned at this point.

devversion committed:
build: increase parallelism for "test_docs_examples" job

mary-poppins commented:
You can preview fab1b76 at https://pr27937-fab1b76.ngbuilds.io/.

devversion commented on scripts/saucelabs/start-tunnel.sh:
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
Sure. Makes sense. Also there are a lot of other scripts in the while repo that use `!/bin/bash`. We should try to switch them over at some point (out of scope for this PR)

devversion commented on .circleci/config.yml:
```diff
@@ -462,3 +513,9 @@ workflows:
+# - don't build the g3 branch
+# - verify that we are bootstrapping with the right yarn version coming from the docker image
+# - check local chrome version pulled from docker image
+# - remove /tools/ngcontainer
```
@gkalpak It's not planned to address this as part of this PR, unless @IgorMinar wants me to do that.

I think tracking this fits better into Jira, so you someone who has access needs to create one. Maybe Alex already has something on this list?

gkalpak replied:
@devversion, sorry for the late response. Are you (or anyone else) working on these or should we create a Jira issue to track?

devversion replied:
@gkalpak No problem 😄 I'm currently not working on these, nor we have these tracked anywhere yet (except in this file).

Agreed, I think it would be good to move these to JIRA. I personally would like to be involved with the removal of `ngcontainer` since we still use it on Material and I've been trying to remove it already in #26691. If you create one, can you mention that Material still depends on it atm?

mary-poppins commented:
You can preview 8b26cc2 at https://pr27937-8b26cc2.ngbuilds.io/.

alexeagle reviewed (approved):
I think developer.md needs a bunch more work, but maybe in a follow-up PR since this is close to landing. 
should just be `yarn build`, `yarn test`, `yarn e2e` and I don't see why ppl want to wait to build the real npm packages, what's the workflow for that?
Let's go through it together?

IgorMinar commented:
We can't remove ngcontainer yet. I talked to AlexE. Too many other repos
need it.

On Sat, Jan 12, 2019, 1:38 PM Paul Gschwendtner <notifications@github.com
wrote:

> *@devversion* commented on this pull request.
> ------------------------------
>
> In .circleci/config.yml
> <https://github.com/angular/angular/pull/27937#discussion_r247326959>:
>
> > @@ -462,3 +513,9 @@ workflows:
>
>              branches:
>
>                only:
>
>                  - master
>
> +
>
> +# TODO:
>
> +# - don't build the g3 branch
>
> +# - verify that we are bootstrapping with the right yarn version coming from the docker image
>
> +# - check local chrome version pulled from docker image
>
> +# - remove /tools/ngcontainer
>
>
> @gkalpak <https://github.com/gkalpak> No problem 😄 I'm currently not
> working on these, nor we have these tracked anywhere yet (except in this
> file).
>
> Agreed, I think it would be good to move these to JIRA. I personally would
> like to be involved with the removal of ngcontainer since we still use it
> on Material and I've been trying to remove it already in #26691
> <https://github.com/angular/angular/pull/26691>. If you create one, can
> you mention that Material still depends on it atm?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/27937#discussion_r247326959>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6FQmBkpSU-HJgUJkCq6e2Ukg_65Eks5vClXngaJpZM4ZskeU>
> .
>
